dissect/hypervisor/disk/hdd.py

```diff
@@ -0,0 +1,392 @@
+
+            sector += read_sectors
+            count -= read_sectors
+            stream_idx += 1
+
+        return b"".join(result)
+
+
+class HDS(AlignedStream):
+    """Parallels HDS implementation.
+
+    HDS is the format for Parallels sparse disk files.
+
+    Args:
+        fh: The file-like object to the HDS file.
+        parent: Optional file-like object for the parent HDS file.
+    """
+
+    def __init__(self, fh: BinaryIO, parent: Optional[BinaryIO] = None):
+        self.fh = fh
+        self.parent = parent
+
+        self.header = c_hdd.pvd_header(fh)
+        if self.header.m_Sig not in (c_hdd.SIGNATURE_STRUCTURED_DISK_V1, c_hdd.SIGNATURE_STRUCTURED_DISK_V2):
+            raise InvalidHeaderError(f"Invalid HDS header signature: {self.header.m_Sig}")
+
+        if self.header.m_Sig == c_hdd.SIGNATURE_STRUCTURED_DISK_V1:
+            size = self.header.m_SizeInSectors_v1
+            self._bat_step = self.header.m_Sectors
+            self._bat_multiplier = 1
+        else:
+            size = self.header.m_SizeInSectors_v2
+            self._bat_step = 1
+            self._bat_multiplier = self.header.m_Sectors
+
+        self.cluster_size = self.header.m_Sectors * SECTOR_SIZE
+
+        self.data_offset = self.header.m_FirstBlockOffset
+        self.in_use = self.header.m_DiskInUse == c_hdd.SIGNATURE_DISK_IN_USE
+
+        super().__init__(size * SECTOR_SIZE)
+
+    @cached_property
+    def bat(self) -> list[int]:
+        """Return the block allocation table (BAT)."""
+        self.fh.seek(len(c_hdd.pvd_header))
+        return c_hdd.uint32[self.header.m_Size](self.fh)
+
+    def _read(self, offset: int, length: int) -> bytes:
+        result = []
+
+        for read_offset, read_size in self._iter_runs(offset, length):
+            # Sentinel value for sparse runs
+            if read_offset is None:
+                if self.parent:
+                    self.parent.seek(offset)
+                    result.append(self.parent.read(read_size))
+                else:
+                    result.append(b"\x00" * read_size)
+            else:
+                self.fh.seek(read_offset)
+                result.append(self.fh.read(read_size))
+
+            offset += read_size
+            length -= read_size
+
+        return b"".join(result)
+
+    def _iter_runs(self, offset: int, length: int) -> Iterator[Tuple[int, int]]:
+        """Iterate optimized read runs for a given offset and read length.
+
+        Args:
+            offset: The offset in bytes to generate runs for.
+            length: The length in bytes to generate runs for.
+        """
+        bat = self.bat
+
+        run_offset = None
+        run_size = 0
+
+        while offset < self.size and length > 0:
+            cluster_idx, offset_in_cluster = divmod(offset, self.cluster_size)
+            read_size = min(self.cluster_size - offset_in_cluster, length)
+
+            bat_entry = bat[cluster_idx]
+            if bat_entry == 0:
```

Comment:
Am I correct in understanding this indicates a sparse cluster? If so, maybe add a comment.

Reply:
Yes, added more comments.